core/src/main/scala/org/apache/spark/memory/MemoryManager.scala

```diff
@@ -255,7 +256,15 @@ private[spark] abstract class MemoryManager(
     }
     val size = ByteArrayMethods.nextPowerOf2(maxTungstenMemory / cores / safetyFactor)
     val default = math.min(maxPageSize, math.max(minPageSize, size))
-    conf.get(BUFFER_PAGESIZE).getOrElse(default)
+    val sizeAsBytes = conf.get(BUFFER_PAGESIZE).getOrElse(default)
+    if (conf.get(MEMORY_ONHEAP_PAGESIZE_OPTIMIZE_ENABLED) &&
+      tungstenMemoryMode == MemoryMode.ON_HEAP &&
+      sizeAsBytes % (1024 * 1024) == 0 &&
+      Utils.isHumongousAllocation(sizeAsBytes)) {
+      sizeAsBytes - Platform.LONG_ARRAY_OFFSET
```

Comment:
what if we always pick this as the page size?

Reply:
I suppose we lose a few bytes in the allocation, and maybe that makes some nice power-of-two data structure not fit, but, I wonder if that's pretty rare and if we can just go with this always indeed

Reply:
you mean even if it is not homongous allocation?

Reply:
Yes. I think the current logic is good, just wondering if it matter much if we do it in all cases. Maybe it's bad for small allocations, but is the offset ever significant relative to the allocation size I wonder? probably not. I wonder if there are future cases, different GCs, that we're not checking here that also need this treatment

Reply:
I revisit here, it's really no need to restrict to only homongous allocation